### ISSUE TITLE
Output immediate store output paths during build

### ISSUE THREAD

**Issue by Cynerd** [NixOS/nix#7369]
**Is your feature request related to a problem? Please describe.**
Add an option to print not just output but also all intermediate outputs used during the build.

**Describe the solution you'd like**
I want option such as `--print-all-paths` that would cause nix to build every store path it uses during the build.

**Describe alternatives you've considered**
The `cachix watch-exec` was considered but not sufficient. Please continue reading to the next section to understand.

**Additional context**
It is possible to use `nix build --no-link --print-out-paths` to get the store path for the build output. This is handy with Cachix as I can do `nix build ... | cahix push foo`. This uploads only store output and everything it depends on, simply because it is unaware of anything else. There is also `cachix watch-exec` that can be used to upload anything that is created as part of the build, but that has an issue if multiple builds are running in parallel (can and will upload stores that are not part of the build to possibly the wrong cache). By printing not only the output path but also the intermediate path, I could use `nix build --no-link --print-all-paths | cachix push foo`.

The reason why I want this is that I use Nix to build my projects with various compilers and for various architectures in CI. By utilizing cache, I can very quickly recreate the CI environment on my PC, but only if cache has all build dependencies, not only build output for the top-level derivation.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
